decompyle3/semantics/transform.py

```diff
@@ -0,0 +1,184 @@
+        return node
+
+        # if key.kind in table:
+        #     self.template_engine(table[key.kind], node)
+        #     self.prune()
+
+    def n_classdef(self, node):
+        self.n_classdef3(node)
+        return node
+
+    def n_classdef3(self, node):
+        # class definition ('class X(A,B,C):')
+
+        # Pick out various needed bits of information
+        # * class_name - the name of the class
+        # * subclass_info - the parameters to the class  e.g.
+        #      class Foo(bar, baz)
+        #               ----------
+        # * subclass_code - the code for the subclass body
+        if node == "classdefdeco2":
+            class_name = node[1][1].attr
+            build_class = node
+        else:
+            build_class = node[0]
+            if build_class == "build_class_kw":
+                mkfunc = build_class[1]
+                assert mkfunc == "mkfunc"
+                if hasattr(mkfunc[0], "attr") and iscode(mkfunc[0].attr):
+                    subclass_code = mkfunc[0].attr
+                else:
+                    assert mkfunc[0] == "load_closure"
+                    subclass_code = mkfunc[1].attr
+                    assert iscode(subclass_code)
+            if build_class[1][0] == "load_closure":
+                code_node = build_class[1][1]
+            else:
+                code_node = build_class[1][0]
+            class_name = code_node.attr.co_name
+
+        assert "mkfunc" == build_class[1]
+        self.currentclass = str(class_name)
+        return node
+
+    def n_ifstmt(self, node):
+        """Here we are just going to check if we can turn an 'ifstmt' into 'assert'"""
+        testexpr = node[0]
+        ifstmts_jump = node[1]
+        if testexpr != "testexpr" or node[1] != "_ifstmts_jump":
+            # No dice
+            return node
+        stmts = ifstmts_jump[0]
+        if stmts in ("c_stmts",) and len(stmts) == 1:
+            stmt = stmts[0]
+            raise_stmt = stmt[0]
+            if raise_stmt == "raise_stmt1" and len(testexpr[0]) == 2:
+                assert_expr = testexpr[0][0]
+                assert_expr.kind = "assert_expr"
+                jmp_true = testexpr[0][1]
+                expr = raise_stmt[0]
+                RAISE_VARARGS_1 = raise_stmt[1]
+                if expr[0] == "call":
+                    # ifstmt
+                    #     0. testexpr
+                    #         testtrue (2)
+                    #             0. expr
+                    #     1. _ifstmts_jump (2)
+                    #         0. c_stmts
+                    #             stmt
+                    #                 raise_stmt1 (2)
+                    #                     0. expr
+                    #                         call (3)
+                    #                     1. RAISE_VARARGS_1
+                    # becomes:
+                    # assert2 ::= assert_expr jmp_true LOAD_ASSERT expr RAISE_VARARGS_1 come_from_pop
```

Comment:
Note that `come_from_pop` mentioned in the comment isn't really generated. But that's okay because it really isn't used in translation to Python source code. 

I'll note however that when you are interested in location determination based on bytecode offset, this however may be of interest.

Reply:
Hmm. There _is_ no `POP_TOP` here. That rule, copied from Python 2.6 is a little different in 3.7 and the comment is not quite right.

@x0ret By the way that `LOAD_ASSERT` is in fact `LOAD_GLOBAL` in 3.7, but perhaps the "ingest" should look for that and change the `LOAD_GLOBAL` to ``LOAD_ASSERT`. 

What do you think?

Reply:
I did a review on "ingest" phase and found out that we should handle 2-arg assert for `LOAD_ASSERT` too. With changes in `scanner3.py` we can detect it now. Let me know if it is not correct.

Reply:
@x0ret note that your changes are now in master and this branch has been rebased off of that.